### PR TITLE
Support `Trilogy` gem

### PR DESCRIPTION
**Pull request by y-yagi**
**What does this PR do?**

This PR adds the support of `Trilogy` gem.

**Motivation:**

The  `Trilogy` gem is MySQL-compatible database adapter and developed by GitHub.
https://github.blog/2022-08-25-introducing-trilogy-a-new-database-adapter-for-ruby-on-rails/

Rails has supported this since v7.1.
https://guides.rubyonrails.org/7_1_release_notes.html#introduce-adapter-for-trilogy

GitHub also has introduced the Active Record database adapter for older Rails. So users can use this gem with Rails before v7.1.
https://github.com/trilogy-libraries/activerecord-trilogy-adapter

I would like to add support for this gem in `ddtrace`.

**Additional Notes:**

`Trilogy` gem doesn't set `required_ruby_version`, so I'm not sure whether it works on older Ruby or not. But Ruby 2.x series are already EOL. So I set Ruby versions for test only 3.x series.

**How to test the change?**

I added test files for this change.

**For Datadog employees:**
- [ ] If this PR touches code that signs or publishes builds or packages, or handles
credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [ ] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: docs/GettingStarted.md [hunk 1]
```diff
@@ -2023,6 +2023,29 @@ end
 LogJob.perform_async('login')
 ```
 
+### Trilogy
+
+The trilogy integration traces any SQL command sent through `trilogy` gem.
```
Comment (estherk15) [inline docs/GettingStarted.md:2028]:
```suggestion
The trilogy integration traces any SQL command sent through the `trilogy` gem.
```

File: docs/GettingStarted.md [hunk 2]
```diff
@@ -2023,6 +2023,29 @@ end
...
+end
+
+client = Trilogy.new(host: "localhost", username: "root")
+client.query("SELECT * FROM users WHERE group='x'")
+```
+
+`options` are the following keyword arguments:
+
+| Key                   | Env Var                        | Description                                                                                                                                                                                                                                                                                                                                                             | Default      |
+|-----------------------|--------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------|
```
Comment (estherk15) [inline docs/GettingStarted.md:2045]:
```suggestion
|-----------------------|---------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|
```

File: docs/GettingStarted.md [hunk 3]
```diff
@@ -2023,6 +2023,29 @@ end
...
+| `service_name`        | `DD_TRACE_TRILOGY_SERVICE_NAME` | Name of application running the `trilogy` instrumentation. May be overridden by `global_default_service_name`. [See *Additional Configuration* for more details](#additional-configuration)                                                                                                                                                                              | `trilogy`     |
```
Comment (estherk15) [inline docs/GettingStarted.md:2046]:
```suggestion
| `service_name`        | `DD_TRACE_TRILOGY_SERVICE_NAME` | Name of application running the `trilogy` instrumentation. May be overridden by `global_default_service_name`. [See *Additional Configuration* for more details](#additional-configuration)  | `trilogy` |
```

File: docs/GettingStarted.md [hunk 4]
```diff
@@ -2023,6 +2023,29 @@ end
...
+| `peer_service`        | `DD_TRACE_TRILOGY_PEER_SERVICE` | Name of external service the application connects to                                                                                                                                                                                                                                                                                                                    | `nil`        |
```
Comment (estherk15) [inline docs/GettingStarted.md:2047]:
```suggestion
| `peer_service`        | `DD_TRACE_TRILOGY_PEER_SERVICE` | Name of external service the application connects to                                                                                                                                         | `nil`     |
```

**Review by estherk15 — APPROVED** [review]
Looks good, left some very optional feedback

**Comment by marcotc** [comment]
Thank you so much, @y-yagi! We'll likely give your changes a thorough review after the U.S. thanksgiving holiday.

**Comment by y-yagi** [comment]
@estherk15 
Thanks for your review! I fixed all your pointed!

**Review by marcotc — APPROVED** [review]
Awesome work, @y-yagi!

One last thing, please take a look at the type checking Action that I just allowed to run, and try to address the typing issues if possible: https://github.com/DataDog/dd-trace-rb/actions/runs/6975162485/job/19386758433?pr=3274
Because third-party integrations heavily integrate with external gems, it might not be possible to fully type it, which is totally fine (and you add a file to the ignore list). The rule of thumb is: if it's typing gets crazy, feel free to ignore the file.

**Review by marcotc — APPROVED** [review]
Thank you so much, @y-yagi! 🙇

**Pull request merged**